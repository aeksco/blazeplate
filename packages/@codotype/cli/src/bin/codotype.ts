#!/usr/bin/env node

// Check node version before requiring/doing anything else
// The user may be on a very old node version
// import minimist from "minimist";
import * as program from "commander";
import chalk from "chalk";
import * as semver from "semver";
import { buildCommand } from "../commands/build";
import { runCommand } from "../commands/run";
import { serveCommand } from "../commands/serve";
import { doctorCommand } from "../commands/doctor";
import { uiCommand } from "../commands/ui";
import { CommandOptions } from "../types";

// TODO - pull this from package.json
const requiredVersion = ">=8.9";

function checkNodeVersion(wanted, id) {
  if (!semver.satisfies(process.version, wanted)) {
    console.log(
      chalk.red(
        "You are using Node " +
          process.version +
          ", but this version of " +
          id +
          " requires Node " +
          wanted +
          ".\nPlease upgrade your Node version."
      )
    );
    process.exit(1);
  }
}

checkNodeVersion(requiredVersion, "codotype-cli");

// // // //

// TODO - this should be pulled from package.json
program.version("1.0.0").usage("<command> [options]");

program
  .command("ui")
  .description("runs the Codotype ui")
  .action((cmd) => {
    const options = cleanArgs(cmd);
    uiCommand(options);
  });

program
  .command("generator-doctor")
  .description("validates the generator in the current working directory")
  .action((cmd) => {
    const options = cleanArgs(cmd);
    doctorCommand(options);
  });

program
  .command("generator-build")
  .description(
    "builds a distributable client app for a single Codotype generator"
  )
  .action((cmd) => {
    const options = cleanArgs(cmd);

    buildCommand(options);
  });

program
  .command("generator-run <project>")
  .option(
    "-p --project <codotype-project.json>",
    "Runs the generator against the codotypeconfiguration file to pass into Codotype runtime"
  )
  // .parse(process.argv)
  .description(
    "run the generator in the current working directory against the blueprint argument"
  )
  .action((project, cmd) => {
    const options = cleanArgs(cmd);
    runCommand(project, options);
  });

program
  .command("generator-serve")
  .description("serves the UI for a single Codotype generator")
  .action((cmd) => {
    const options = cleanArgs(cmd);
    serveCommand(options);
  });

// output help information on unknown commands
program.arguments("<command>").action((cmd) => {
  program.outputHelp();
  console.log(`  ` + chalk.red(`Unknown command ${chalk.yellow(cmd)}.`));
  console.log();
});

// add some useful info on help
program.on("--help", () => {
  console.log();
  console.log(
    `  Run ${chalk.cyan(
      `codotype <command> --help`
    )} for detailed usage of given command.`
  );
  console.log();
});

program.commands.forEach((c) => c.on("--help", () => console.log()));

// Parse arguments into commander program
program.parse(process.argv);

if (!process.argv.slice(2).length) {
  program.outputHelp();
}

/**
 * camelize
 * @param str - string that's being camelized
 */
function camelize(str: string): string {
  return str.replace(/-(\w)/g, (_, c) => (c ? c.toUpperCase() : ""));
}

// commander passes the Command object itself as options,
// extract only actual options into a fresh object.
function cleanArgs(cmd): CommandOptions {
  const args = {};
  cmd.options.forEach((o) => {
    const key: string = camelize(o.long.replace(/^--/, ""));
    // if an option is not present and Command has a method with the same name
    // it should not be copied
    if (typeof cmd[key] !== "function" && typeof cmd[key] !== "undefined") {
      args[key] = cmd[key];
    }
  });
  return args;
}
