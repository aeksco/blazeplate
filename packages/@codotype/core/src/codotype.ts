import { ConfigurationGroupPropertyBuilder } from "./factories/ConfigurationGroupProperty";
import { SchemaBuilder } from "./factories/Schema";
import { ConfigurationGroupBuilder } from "./factories/ConfigurationGroup";
import { ConfigurationGroupSectionBuilder } from "./factories/ConfigurationGroupSection";
import { SchemaEditorBuilder } from "./factories/SchemaEditor";
import { RelationBuilder } from "./factories/Relation";
import { AttributeBuilder } from "./factories/Attribute";

// // // //

export interface CodotypeFactory {
    ConfigurationGroup: typeof ConfigurationGroupBuilder;
    ConfigurationGroupSection: typeof ConfigurationGroupSectionBuilder;
    ConfigurationGroupProperty: typeof ConfigurationGroupPropertyBuilder;
    Attribute: typeof AttributeBuilder;
    Relation: typeof RelationBuilder;
    Schema: typeof SchemaBuilder;
    SchemaEditor: typeof SchemaEditorBuilder;
    // Generator: typeof GeneratorBuilder; // TODO - add GeneratorBuilder
    // Plugin: typeof PluginBuilder; // TODO - add PluginBuilder
}

// TODO - abstract into @codotype/primatives
export const Codotype: CodotypeFactory = {
    ConfigurationGroup: ConfigurationGroupBuilder,
    ConfigurationGroupSection: ConfigurationGroupSectionBuilder,
    ConfigurationGroupProperty: ConfigurationGroupPropertyBuilder,
    Attribute: AttributeBuilder,
    Relation: RelationBuilder,
    Schema: SchemaBuilder,
    SchemaEditor: SchemaEditorBuilder,
};
