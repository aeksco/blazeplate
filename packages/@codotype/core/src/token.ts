/**
 * TokenPluralization
 * Encapsulates both singular and plural TokenCasing objects - @see Schema.identifiers
 * @param singular - the singular TokenCasing object
 * @param plural - the plural TokenCasing object
 */
export interface TokenPluralization {
    singular: TokenCasing;
    plural: TokenCasing;
}

/**
 * TokenCasing
 * Encapsulates different tokens used by a Plugin to produce code that matches situational expectations
 * @param label - the title-cased token, i.e. `My Project`
 * @param snake - the snake-cased token, i.e. `my_project`
 * @param camel - the camel-cased token, i.e. `myProject`
 * @param pascal - the pascal-cased token, i.e. `MyProject`
 * @param kebab - the kebab-cased token, i.e. `my-project`
 */
export interface TokenCasing {
    label: string;
    snake: string;
    camel: string;
    pascal: string;
    kebab: string;
}

/**
 * EMPTY_TOKEN_CASING
 * Defines an empty TokenCasing object
 */
export const EMPTY_TOKEN_CASING: TokenCasing = {
    label: "",
    snake: "",
    camel: "",
    pascal: "",
    kebab: "",
};
