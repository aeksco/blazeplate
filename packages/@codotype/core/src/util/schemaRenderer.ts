import { SchemaInput, Datatype, Attribute, Schema, Relation } from "../";
import { inflateSchema, buildRelations } from "./inflate";
import { ProjectInput } from "../project";

// // // //

// CLEANUP - document this function, write better tests
export const getDatatypeValueJson = ({
    datatype,
}: {
    datatype: Datatype;
}): string => {
    switch (datatype) {
        // case Datatype.ID:
        //   return `1`;
        case Datatype.UUID:
            return `"00112233-4455-6677-8899-aabbccddeeff"`;
        case Datatype.STRING:
            return `"string"`;
        case Datatype.TEXT:
            return `"longer string for text"`;
        case Datatype.INT:
            return `128`;
        case Datatype.BIGINT:
            return `12378971290123987`;
        case Datatype.FLOAT:
            return `3.14159`;
        case Datatype.DECIMAL:
            return `3.14159`;
        case Datatype.NUMERIC:
            return `3.14159`;
        case Datatype.BOOLEAN:
            return `true`;
        case Datatype.JSON:
            return `{}`;
        case Datatype.JSONB:
            return `{}`;
        case Datatype.DATE:
            return `"2019-03-11"`;
        case Datatype.TIME:
            return `"17:04:14 GMT-0400"`;
        case Datatype.DATETIME:
            return `"3/18/2019, 5:04:51 PM"`;
        case Datatype.TIMESTAMP:
            return `"3/18/2019, 5:04:51 PM"`;
        case Datatype.UUID_ARRAY:
            return `["00112233-4455-6677-8899-aabbccddeeff"]`;
        case Datatype.STRING_ARRAY:
            return `["string", "array"]`;
        case Datatype.TEXT_ARRAY:
            return `["text", "array"]`;
        case Datatype.INT_ARRAY:
            return `[128, 256]`;
        case Datatype.BIGINT_ARRAY:
            return `[128128309810198, 128128309810198]`;
        case Datatype.FLOAT_ARRAY:
            return `[3.14156, 64.23012]`;
        case Datatype.DECIMAL_ARRAY:
            return `[0.12390123, 0.12312442]`;
        case Datatype.NUMERIC_ARRAY:
            return `[128, 255.9]`;
        case Datatype.BOOLEAN_ARRAY:
            return `[true, false, true]`;
        case Datatype.DATE_ARRAY:
            return `["2019-03-11", "2020-03-11]`;
        case Datatype.TIME_ARRAY:
            return `["17:04:14 GMT-0400", "12:04:14 GMT-0400"]`;
        case Datatype.DATETIME_ARRAY:
            return `[128]`;
        case Datatype.TIMESTAMP_ARRAY:
            return `["3/18/2019, 5:04:51 PM", "3/20/2019, 1:04:51 PM"]`;
    }
};

export function renderSchemaJson({
    schemaInput,
    projectInput,
}: {
    projectInput: ProjectInput;
    schemaInput: SchemaInput;
}): string {
    const inflatedSchema: Schema = inflateSchema({
        schemaInput,
        relations: buildRelations({
            schemaInputs: projectInput.schemas,
            relationInputs: projectInput.relations,
        }),
    });
    // Define + open JSON output
    let jsonOutput: string[] = [
        "{", // Open JSON output
        // Map each property
        [
            ...schemaInput.attributes.map((attr: Attribute) => {
                return `  "${attr.identifiers.snake}": ${getDatatypeValueJson({
                    datatype: attr.datatype,
                })}`;
            }),
            ...inflatedSchema.relations.map((r: Relation): string => {
                return `  "${r.identifiers.destination.alias.singular.camel}": "${r.identifiers.destination.canonical.singular.pascal} ID"`;
            }),
        ].join(",\n"),
        "}", // Close JSON output
    ];

    // Joins lines
    return jsonOutput.join("\n");
}

// // // //

// CLEANUP - document this function, write better tests
export const getDatatypeValueGraphQL = ({
    datatype,
}: {
    datatype: Datatype;
}): string => {
    switch (datatype) {
        case Datatype.UUID:
            return `String`;
        case Datatype.STRING:
            return `String`;
        case Datatype.TEXT:
            return `String`;
        case Datatype.INT:
            return `Number`;
        case Datatype.BIGINT:
            return `"Number`;
        case Datatype.FLOAT:
            return `Number`;
        case Datatype.DECIMAL:
            return `Number`;
        case Datatype.NUMERIC:
            return `Number`;
        case Datatype.BOOLEAN:
            return `Boolean`;
        case Datatype.JSON:
            return `"{}"`;
        case Datatype.JSONB:
            return `"{}"`;
        case Datatype.DATE:
            return `Date`;
        case Datatype.TIME:
            return `Time`;
        case Datatype.DATETIME:
            return `Date`;
        case Datatype.TIMESTAMP:
            return `Date`;
        case Datatype.UUID_ARRAY:
            return `[String]`;
        case Datatype.STRING_ARRAY:
            return `[String]`;
        case Datatype.TEXT_ARRAY:
            return `[String]`;
        case Datatype.INT_ARRAY:
            return `[Number]`;
        case Datatype.BIGINT_ARRAY:
            return `[Number]`;
        case Datatype.FLOAT_ARRAY:
            return `[Number]`;
        case Datatype.DECIMAL_ARRAY:
            return `[Number]`;
        case Datatype.NUMERIC_ARRAY:
            return `[Number]`;
        case Datatype.BOOLEAN_ARRAY:
            return `[Boolean]`;
        case Datatype.DATE_ARRAY:
            return `[Date]`;
        case Datatype.TIME_ARRAY:
            return `[Time]`;
        case Datatype.DATETIME_ARRAY:
            return `[Date]`;
        case Datatype.TIMESTAMP_ARRAY:
            return `[Date]`;
    }
};

export function renderSchemaGrapqhQL({
    schemaInput,
    projectInput,
}: {
    projectInput: ProjectInput;
    schemaInput: SchemaInput;
}): string {
    const inflatedSchema: Schema = inflateSchema({
        schemaInput,
        relations: buildRelations({
            schemaInputs: projectInput.schemas,
            relationInputs: projectInput.relations,
        }),
    });
    // Define + open JSON output
    let jsonOutput: string[] = [
        `type ${schemaInput.identifiers.singular.pascal} {`, // Open JSON output
        // Map each property
        ...schemaInput.attributes.map((attr: Attribute) => {
            return `  ${attr.identifiers.snake}: ${getDatatypeValueGraphQL({
                datatype: attr.datatype,
            })}!`;
        }),
        ...inflatedSchema.relations.map((r: Relation): string => {
            return `  ${r.identifiers.destination.alias.singular.camel}: ${r.identifiers.destination.canonical.singular.pascal}!`;
        }),
        "}", // Close JSON output
    ];

    // Joins lines
    return jsonOutput.join("\n");
}

// // // //

// CLEANUP - document this function, write better tests
export const getDatatypeValueTypeScript = ({
    datatype,
}: {
    datatype: Datatype;
}): string => {
    switch (datatype) {
        case Datatype.UUID:
            return `string`;
        case Datatype.STRING:
            return `string`;
        case Datatype.TEXT:
            return `string`;
        case Datatype.INT:
            return `number`;
        case Datatype.BIGINT:
            return `"number`;
        case Datatype.FLOAT:
            return `number`;
        case Datatype.DECIMAL:
            return `number`;
        case Datatype.NUMERIC:
            return `number`;
        case Datatype.BOOLEAN:
            return `boolean`;
        case Datatype.JSON:
            return `"{}"`;
        case Datatype.JSONB:
            return `"{}"`;
        case Datatype.DATE:
            return `string`;
        case Datatype.TIME:
            return `string`;
        case Datatype.DATETIME:
            return `string`;
        case Datatype.TIMESTAMP:
            return `string`;
        case Datatype.UUID_ARRAY:
            return `string[]`;
        case Datatype.STRING_ARRAY:
            return `string[]`;
        case Datatype.TEXT_ARRAY:
            return `string[]`;
        case Datatype.INT_ARRAY:
            return `number[]`;
        case Datatype.BIGINT_ARRAY:
            return `number[]`;
        case Datatype.FLOAT_ARRAY:
            return `number[]`;
        case Datatype.DECIMAL_ARRAY:
            return `number[]`;
        case Datatype.NUMERIC_ARRAY:
            return `number[]`;
        case Datatype.BOOLEAN_ARRAY:
            return `boolean[]`;
        case Datatype.DATE_ARRAY:
            return `string[]`;
        case Datatype.TIME_ARRAY:
            return `string[]`;
        case Datatype.DATETIME_ARRAY:
            return `string[]`;
        case Datatype.TIMESTAMP_ARRAY:
            return `string[]`;
    }
};

export function renderSchemaTypeScript({
    schemaInput,
    projectInput,
}: {
    projectInput: ProjectInput;
    schemaInput: SchemaInput;
}): string {
    const inflatedSchema: Schema = inflateSchema({
        schemaInput,
        relations: buildRelations({
            schemaInputs: projectInput.schemas,
            relationInputs: projectInput.relations,
        }),
    });
    // Define + open JSON output
    let output: string[] = [
        `interface ${schemaInput.identifiers.singular.pascal} {`, // Open JSON output
        // Map each property
        ...schemaInput.attributes.map((attr: Attribute) => {
            return `  ${attr.identifiers.snake}: ${getDatatypeValueTypeScript({
                datatype: attr.datatype,
            })};`;
        }),
        ...inflatedSchema.relations.map((r: Relation): string => {
            return `  ${r.identifiers.destination.alias.singular.camel}Id: ${r.identifiers.destination.canonical.singular.pascal};`;
        }),
        "}", // Close JSON output
    ];

    // Joins lines
    return output.join("\n");
}
