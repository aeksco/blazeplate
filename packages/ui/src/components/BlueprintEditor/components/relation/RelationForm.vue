<template>
  <b-modal
    lazy
    :visible="showingModal"
    :title="'New Relation'"
    ok-title='Submit'
    ok-variant='success'
    cancel-title='Cancel'
    cancel-variant='light'
    @ok="submit(newModel)"
    @hide="showModal(false)"
  >

    <b-form>
      <p class="mb-2 form-text text-muted">TODO - add relation form here</p>
      <input type="" v-model="newModel.label" name="">
    </b-form>

  </b-modal>
</template>

<script>
import { mapGetters, mapActions, mapMutations } from 'vuex'

export default {
  name: 'RelationForm',
  computed: mapGetters({
    showingModal: 'editor/schema/relation/modals/form/showing',
    newModel: 'editor/schema/relation/collection/newModel'
  }),
  methods: {
    ...mapActions({
      createPersonModel: 'editor/schema/relation/collection/create',
      updateParentSchemaRelations: 'editor/schema/updateRelations'
    }),
    ...mapMutations({
      showModal: 'editor/schema/relation/modals/form/showing',
      setNewPersonModel: 'editor/schema/relation/collection/newModel'
    }),
    submit (newModel) {
      this.setNewPersonModel(newModel)
      this.createPersonModel(),
      this.updateParentSchemaRelations()
    }
  }
}
</script>
