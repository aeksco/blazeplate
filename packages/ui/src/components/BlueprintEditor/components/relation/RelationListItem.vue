<template>
  <li class='list-group-item'>
    {{ model }}
  </li>
</template>

<script>
export default {
  name: 'AttributeListItem',
  props: ['model']
}
</script>
