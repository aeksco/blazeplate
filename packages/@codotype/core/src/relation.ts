import { UUID } from "./uuid";
import { TokenPluralization } from "./token";
import { SchemaCreators } from "./schema";

// // // //

/**
 * RelationType
 * Defines an enum of available RelationsTypes
 */
export type RelationType =
    | "BELONGS_TO"
    | "HAS_ONE"
    | "HAS_MANY"
    | "TO_ONE"
    | "TO_MANY"
    | "HAS_AND_BELONGS_TO_MANY";
export enum RelationTypes {
    BELONGS_TO = "BELONGS_TO",
    HAS_ONE = "HAS_ONE",
    HAS_MANY = "HAS_MANY",
    TO_ONE = "TO_ONE",
    TO_MANY = "TO_MANY",
    HAS_AND_BELONGS_TO_MANY = "HAS_AND_BELONGS_TO_MANY",
}

/**
 * RelationInput
 * Defines the RelationInput interface
 * TODO - add support for addons here -> see AttributeAddon for context
 */
export interface RelationInput {
    id: UUID;
    type: RelationType;
    required: boolean; // TODO - keep this for now - should this be handled as an Addon? YES.
    source: SchemaCreators;
    sourceSchemaID: UUID;
    destinationSchemaID: UUID;
    sourceSchemaAlias: string;
    destinationSchemaAlias: string;
}

// Relation constants
export const DEFAULT_RELATION_INPUT: RelationInput = {
    id: "",
    type: null,
    required: false,
    source: SchemaCreators.user,
    sourceSchemaID: "",
    destinationSchemaID: "",
    sourceSchemaAlias: "",
    destinationSchemaAlias: "",
};

// // // //

/**
 * Relation
 * Defines the Reference interface
 * TODO - add support for addons here -> see AttributeAddon for context
 */
export interface Relation {
    id: UUID;
    type: RelationType;
    sourceRelationId: UUID;
    sourceSchemaId: UUID;
    destinationSchemaId: UUID;
    identifiers: {
        source: {
            canonical: TokenPluralization;
            alias: TokenPluralization;
        };
        destination: {
            canonical: TokenPluralization;
            alias: TokenPluralization;
        };
    };
}
