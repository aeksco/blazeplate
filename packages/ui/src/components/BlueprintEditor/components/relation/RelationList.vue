<template>
  <b-row>
    <b-col lg=12>

      <div class='card' id='relation-detail'>

        <div class="card-header d-flex justify-content-between align-items-center">

          <span>
            <i class="fa fa-link mr-1"></i>
            Relationships
          </span>

          <NewModalButton
            id="add-relation-button"
            vuexModule="editor/schema/relation/modals/form/showing"
          />

        </div>

        <b-list-group flush v-if="collection[0]">
          <RelationListItem v-for="item in collection" :key="item.id" :model="item" />
        </b-list-group>

        <!-- NOTES FROM OLD VERSION -->
        <!-- <ul v-if="!relations.length && !referencedBy.length"> -->
        <b-list-group flush v-else>
          <b-list-group-item class="text-center text-primary">
            <i class="fa fa-lg fa-info-circle"></i>
            <p class="mb-0 mt-1">
              <!-- TODO - Click to add your first Relation -->
              Relations define associations between different models
            </p>
          </b-list-group-item>
        </b-list-group>

      </div>

    </b-col>
  </b-row>
</template>

<script>
import { mapGetters } from 'vuex'
import NewModalButton from '../NewModalButton'
import RelationListItem from './RelationListItem'

export default {
  name: 'RelationList',
  components: {
    NewModalButton,
    RelationListItem
  },
  computed: mapGetters({
    collection: 'editor/schema/relation/collection/items'
  })
}
</script>
