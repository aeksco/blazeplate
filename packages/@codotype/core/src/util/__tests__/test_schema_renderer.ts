import { renderSchemaJson } from "../schemaRenderer";
import { testState } from "../../";
import { Codotype } from "../../codotype";
import { EMPTY_TOKEN_CASING } from "../../token";
import { RelationTypes } from "../../relation";

// // // //

const { userSchema, movieSchema } = testState;

describe("/src/schemaRenderer", () => {
    it("renders JSON representation correctly", () => {
        const renderedSchema: string = renderSchemaJson({
            schemaInput: userSchema,
            projectInput: {
                id: "my-project",
                schemas: [userSchema, movieSchema],
                pluginID: "",
                pluginVersion: "",
                configuration: {},
                relations: [
                    new Codotype.Relation({
                        type: RelationTypes.BELONGS_TO,
                        sourceSchemaID: userSchema.id,
                        destinationSchemaID: movieSchema.id,
                    }),
                ],
                identifiers: {
                    ...EMPTY_TOKEN_CASING,
                },
            },
        });
        expect(renderedSchema).toMatchSnapshot();
    });
});
