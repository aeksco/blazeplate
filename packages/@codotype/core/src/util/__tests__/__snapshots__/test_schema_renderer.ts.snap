// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/src/schemaRenderer renders JSON representation correctly 1`] = `
"{
  \\"movie\\": \\"Movie ID\\"
}"
`;
